rust/otap-dataflow/crates/engine/src/retry_processor.rs

```diff
@@ -0,0 +1,692 @@
+            max_pending_messages: 10000,
+            cleanup_interval_secs: 60,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+struct PendingMessage<PData> {
+    data: PData,
+    retry_count: usize,
+    next_retry_time: Instant,
+    last_error: String,
+}
+
+/// A processor that handles message retries with exponential backoff
+///
+/// The RetryProcessor maintains a queue of messages that have failed processing
+/// and retries them according to the configured retry policy. It tracks each
+/// message with a unique ID and implements exponential backoff for retry delays.
+pub struct RetryProcessor<PData: Clone + Send + 'static> {
+    config: RetryConfig,
+    pending_messages: HashMap<u64, PendingMessage<PData>>,
+    next_message_id: u64,
+    last_cleanup_time: Instant,
+}
+
+impl<PData: Clone + Send + 'static> RetryProcessor<PData> {
+    /// Creates a new RetryProcessor with default configuration
+    #[must_use]
+    pub fn new() -> Self {
+        Self::with_config(RetryConfig::default())
+    }
+
+    /// Creates a new RetryProcessor with the specified configuration
+    #[must_use]
+    pub fn with_config(config: RetryConfig) -> Self {
+        Self {
+            config,
+            pending_messages: HashMap::new(),
+            next_message_id: 1,
+            last_cleanup_time: Instant::now(),
+        }
+    }
+
+    fn acknowledge(&mut self, id: u64) {
+        if let Some(_removed) = self.pending_messages.remove(&id) {
+            log::debug!("Acknowledged and removed message with ID: {id}");
+        } else {
+            log::warn!("Attempted to acknowledge non-existent message with ID: {id}");
+        }
+    }
+
+    async fn handle_nack(
+        &mut self,
+        id: u64,
+        reason: String,
+        _effect_handler: &mut EffectHandler<PData>,
+    ) -> Result<(), Error<PData>> {
+        if let Some(mut pending) = self.pending_messages.remove(&id) {
+            pending.retry_count += 1;
+            pending.last_error = reason;
+
+            if pending.retry_count <= self.config.max_retries {
+                let delay_ms = (self.config.initial_retry_delay_ms as f64
+                    * self
+                        .config
+                        .backoff_multiplier
+                        .powi(pending.retry_count as i32 - 1))
+                .min(self.config.max_retry_delay_ms as f64) as u64;
+
+                pending.next_retry_time = Instant::now() + Duration::from_millis(delay_ms);
+                let retry_count = pending.retry_count;
+                let _previous = self.pending_messages.insert(id, pending);
+                log::debug!("Scheduled message {id} for retry attempt {retry_count}");
+            } else {
+                log::error!(
+                    "Message {} exceeded max retries ({}), dropping. Last error: {}",
+                    id,
+                    self.config.max_retries,
+                    pending.last_error
+                );
+            }
+        }
```

Comment:
nit: I noticed that in the `acknowledge` method, we warn if the map doesn't contain the `id`. Should we do the same thing here?
```suggestion
        } else {
            log::warn!("Attempted to handle nack for non-existent message with ID: {id}");
        }
```